WooCommerce/Classes/ViewRelated/Products/Edit Product/EditCategories/ProductCategoryListViewController.swift

```diff
@@ -67,8 +68,18 @@ private extension ProductCategoryListViewController {
 private extension ProductCategoryListViewController {
     func configureViewModel() {
         viewModel.performInitialFetch()
-        viewModel.observeCategoryListChanges { [weak self] in
-            self?.tableView.reloadData()
+        viewModel.observeCategoryListStateChanges { [weak self] syncState in
+            switch syncState {
+            case .syncing:
+                self?.displayPlaceholderCategories()
+            case let .failed(pageNumber):
```

Comment:
I just thought of a possibly **over-engineered approach** to this. I find it odd that the UI knows how to "retry" a failure. I feel like that should just be hidden inside the `ViewModel`. 

I wonder if can have something like:

```swift
struct RetryToken {
    // private so the "view" will not be able to access it
    private let fromPageNumber: Int 
}
```

And we use that instead of `pageNumber`

```swift
case failed(retryToken: RetryToken)
```

And that's what we need to pass when retrying:

```swift
    func retryCategorySynchronization(token: RetryToken) {
        let pageNumber = token.pageNumber
```

Just thinking out loud. I'm not even sure if this works. 😄

Reply:
I'll play with it! It sounds like it should work.

Reply:
https://github.com/woocommerce/woocommerce-ios/pull/2130/commits/36002f6078b23eaa20a2fb81b93a31e5703fb0af